### PR TITLE
Polygon collinear points

### PR DESCRIPTION
**Pull request by skemaikin**

### CODE REVIEW

**Comment by bertt** [comment]
Hi, thanks for the pr. 

Some remarks: 

- it seems the unit AvoidCollinearPoints() fails (and also the existing test DoNotGetNegativeNormal) - see github action run

- In the added unit test I'm missing the purpose of this pull request. Is it a better normal calculation? (btw I like that it doesn't use the epsilon magic number). I see the same number of triangles is generated in the unit test (as compared to the current implementation - 10 triangles).

**Comment by skemaikin** [comment]
> Hi, thanks for the pr.
> 
> Some remarks:
> 
> * it seems the unit AvoidCollinearPoints() fails (and also the existing test DoNotGetNegativeNormal) - see github action run
> * In the added unit test I'm missing the purpose of this pull request. Is it a better normal calculation? (btw I like that it doesn't use the epsilon magic number). I see the same number of triangles is generated in the unit test (as compared to the current implementation - 10 triangles).

It seems to me I understand what's going on. I just copied the `DoNotGetNegativeNormal()` and didn't check it. The call: `var polygon = (Polygon)Geometry.Deserialize<WktSerializer>(wkt);` is incorrect. It needs to be the: `var polygon = (Polygon)Geometry.Deserialize<WktSerializer>(Encoding.UTF8.GetBytes(wkt));`.

**Comment by skemaikin** [comment]
> * In the added unit test I'm missing the purpose of this pull request. Is it a better normal calculation? (btw I like that it doesn't use the epsilon magic number). I see the same number of triangles is generated in the unit test (as compared to the current implementation - 10 triangles).

My normal calculation is really better. I have a large dataset of non-triangulated polygons with the collinear verticies on edges. This corrected code improve the generated picture.

**Comment by bertt** [comment]
> My normal calculation is really better. I have a large dataset of non-triangulated polygons with the collinear verticies on edges. This corrected code improve the generated picture.

Can you add a minimal sample of that dataset (in a unit test) that proves this method is better?

**Comment by bertt** [comment]
btw: keep in mind it's on the roadmap we use another library for making the triangles (one that supports interior rings)

File: src/triangulator.tests/NormalTests.cs [hunk 1]
```diff
@@ -9,13 +11,26 @@ public void DoNotGetNegativeNormal()
...
         var wkt = "POLYGON Z ((-75.51889015799998 39.15624977500005 17.71,-75.51880883299998 39.15605855000007 17.71,-75.51882902999995 39.156053351000025 17.71,-75.51874978099994 39.155858297000066 17.71,-75.51882395499996 39.155839280000066 17.71,-75.51889671399994 39.15601041700006 17.71,-75.51894566199996 39.15599799200004 17.71,-75.51903664399998 39.15621200700008 17.71,-75.51889015799998 39.15624977500005 17.71))";
-        var polygon = (Polygon)Geometry.Deserialize<WktSerializer>(wkt);
+        var polygon = (Polygon)Geometry.Deserialize<WktSerializer>(Encoding.UTF8.GetBytes(wkt));
 
         // act
         var normal = polygon.GetNormal();
 
         // assert
-        Assert.That(normal.X==0 && normal.Y==0 && normal.Z == 0);
+        Assert.That(normal.X==0 && normal.Y==0 && normal.Z == 0, Is.False);
```
Comment (bertt) [inline src/triangulator.tests/NormalTests.cs:20]: do check the real values, not what's is not. I think the z = -1 with this polygon? if so, the purpose was to make to make it not negative, we have to check what the expected normal is for the given polygon.

**Pull request closed without merging**